<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>ganon2</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "ganon2";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> ganon2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href=".">ganon2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#features">Features</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#installation-with-conda">Installation with conda</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#installation-from-source">Installation from source</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#python-dependencies">Python dependencies</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#c-dependencies">C++ dependencies</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#downloading-and-building-ganon-submodules">Downloading and building ganon + submodules</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#installing-raptor">Installing raptor</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#dependencies">Dependencies</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#downloading-and-building-raptor-submodules">Downloading and building raptor + submodules</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#testing">Testing</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#running-tests">Running tests</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#parameters">Parameters</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="start/">Quick Start</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="default_databases/">Databases (ganon build)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="custom_databases/">Custom databases (ganon build-custom)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="classification/">Classification (ganon classify)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="reports/">Reports (ganon report)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="table/">Table (ganon table)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="outputfiles/">Output files</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="tutorials/">Tutorials</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">ganon2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li class="breadcrumb-item active">ganon2</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="ganon">ganon <a href="https://github.com/pirovc/ganon"><img alt="GitHub release (latest by date)" src="https://img.shields.io/github/v/release/pirovc/ganon" /></a><a class="headerlink" href="#ganon" title="Permanent link">:)</a></h1>
<p><a href="https://travis-ci.com/pirovc/ganon"><img alt="Build Status" src="https://travis-ci.com/pirovc/ganon.svg?branch=master" /></a> <a href="https://codecov.io/gh/pirovc/ganon"><img alt="codecov" src="https://codecov.io/gh/pirovc/ganon/branch/master/graph/badge.svg" /></a> <a href="https://anaconda.org/bioconda/ganon"><img alt="Anaconda-Server Badge" src="https://anaconda.org/bioconda/ganon/badges/downloads.svg" /></a> <a href="https://anaconda.org/bioconda/ganon"><img alt="Anaconda-Server Badge" src="https://anaconda.org/bioconda/ganon/badges/platforms.svg" /></a> <a href="http://bioconda.github.io/recipes/ganon/README.html"><img alt="install with bioconda" src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" /></a> <a href="https://dx.doi.org/10.1093/bioinformatics/btaa458"><img alt="Publication" src="https://img.shields.io/badge/DOI-10.1101%2F406017-blue" /></a> </p>
<p>Code: <a href="https://github.com/pirovc/ganon">GitHub repository</a></p>
<p>ganon is designed to index large sets of genomic reference sequences and to classify reads against them efficiently. The tool uses <a href="https://doi.org/10.1186/s13059-023-02971-4">Hierarchical Interleaved Bloom Filters</a> as indices based on k-mers with optional minimizers. It was mainly developed, but not limited, to the metagenomics classification problem: quickly assign sequence fragments to their closest reference among thousands of references. After classification, taxonomic or sequence abundances are estimated and reported.</p>
<h2 id="features">Features<a class="headerlink" href="#features" title="Permanent link">:)</a></h2>
<ul>
<li>integrated download and build of any subset from <a href="default_databases/#refseq-and-genbank">RefSeq/Genbank/GTDB</a> with incremental <a href="default_databases/#update-ganon-update">updates</a></li>
<li>NCBI and <a href="default_databases/#gtdb">GTDB</a> native support for taxonomic classification, custom taxonomy or no taxonomy at all</li>
<li><a href="custom_databases/">customizable database</a> build for local or non-standard sequence files</li>
<li>optimized <a href="classification/#binning">taxonomic binning</a> and <a href="classification/#profiling">profiling</a> configurations</li>
<li>build and classify at various taxonomic levels, strain, assembly, file, sequence or custom specialization</li>
<li><a href="classification/#multiple-and-hierarchical-classification">hierarchical classification</a> using several databases in one or more levels in just one run</li>
<li><a href="classification/#reads-with-multiple-matches">EM and/or LCA</a> algorithms to solve multiple-matching reads</li>
<li>reporting of multiple and unique matches for every read</li>
<li><a href="reports/#report-type-report-type">reporting</a> of sequence, taxonomic or multi-match abundances with optional genome size correction</li>
<li>advanced tree-like <a href="reports">reports</a> with several filter options</li>
<li>generation of <a href="table/">contingency tables</a> with several filters for multi-sample studies</li>
</ul>
<p>ganon achieved very good results in <a href="https://dx.doi.org/10.1093/bioinformatics/btaa458">our own evaluations</a> but also in independent evaluations: <a href="https://lemmi-v1.ezlab.org/">LEMMI</a>, <a href="https://lemmi.ezlab.org/">LEMMI v2</a> and <a href="https://dx.doi.org/10.1038/s41592-022-01431-4">CAMI2</a></p>
<h2 id="installation-with-conda">Installation with conda<a class="headerlink" href="#installation-with-conda" title="Permanent link">:)</a></h2>
<p>The easiest way to install ganon is via conda, using the bioconda and conda-forge channels:</p>
<pre><code class="language-bash">conda install -c bioconda -c conda-forge ganon
</code></pre>
<p>However, there are possible performance benefits compiling ganon from source in the target machine rather than using the conda version. To do so, please follow the instructions below:</p>
<h2 id="installation-from-source">Installation from source<a class="headerlink" href="#installation-from-source" title="Permanent link">:)</a></h2>
<h3 id="python-dependencies">Python dependencies<a class="headerlink" href="#python-dependencies" title="Permanent link">:)</a></h3>
<ul>
<li>python &gt;=3.6</li>
<li>pandas &gt;=1.1.0</li>
<li><a href="https://github.com/pirovc/multitax">multitax</a> &gt;=1.3.1</li>
</ul>
<pre><code class="language-bash"># Python version should be &gt;=3.6
python3 -V

# Install packages via pip or conda:
# PIP
python3 -m pip install &quot;pandas&gt;=1.1.0&quot; &quot;multitax&gt;=1.3.1&quot;
# Conda (alternative)
conda install &quot;pandas&gt;=1.1.0&quot; &quot;multitax&gt;=1.3.1&quot;
</code></pre>
<h3 id="c-dependencies">C++ dependencies<a class="headerlink" href="#c-dependencies" title="Permanent link">:)</a></h3>
<ul>
<li>GCC &gt;=11</li>
<li>CMake &gt;=3.4</li>
<li>zlib</li>
<li>bzip2</li>
<li>raptor &gt;=3.0.1</li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If your system has GCC version 10 or below, you can create an environment with the latest conda-forge GCC version and dependencies: <code>conda create -c conda-forge -n gcc-conda gcc gxx zlib bzip2 cmake</code> and activate the environment with: <code>source activate gcc-conda</code>.</p>
<p>In CMake, you may have set the environment include directory with the following parameter: <code>-DSEQAN3_CXX_FLAGS="-I/path/to/miniconda3/envs/gcc-conda/include/"</code> changing <code>/path/to/miniconda3</code> with your local path to the conda installation.</p>
</div>
<h3 id="downloading-and-building-ganon-submodules">Downloading and building ganon + submodules<a class="headerlink" href="#downloading-and-building-ganon-submodules" title="Permanent link">:)</a></h3>
<pre><code class="language-bash">git clone --recurse-submodules https://github.com/pirovc/ganon.git
</code></pre>
<pre><code class="language-bash"># Install Python side
cd ganon
python3 setup.py install --record files.txt  # optional

# Compile and install C++ side
mkdir -p build
cd build
cmake -DCMAKE_BUILD_TYPE=Release -DVERBOSE_CONFIG=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCONDA=OFF -DLONGREADS=OFF ..
make -j 4
sudo make install  # optional
</code></pre>
<ul>
<li>to change install location (e.g. <code>/myprefix/bin/</code>), set the installation prefix in the cmake command with <code>-DCMAKE_INSTALL_PREFIX=/myprefix/</code></li>
<li>use <code>-DINCLUDE_DIRS</code> to set alternative paths to cxxopts and Catch2 libs.</li>
<li>to classify extremely large reads or contigs that would need more than 65000 k-mers, use <code>-DLONGREADS=ON</code></li>
</ul>
<h3 id="installing-raptor">Installing raptor<a class="headerlink" href="#installing-raptor" title="Permanent link">:)</a></h3>
<p>The easiest way to install <a href="https://github.com/seqan/raptor">raptor</a> is via conda with <code>conda install -c bioconda -c conda-forge "raptor&gt;=3.0.1"</code> (already included in ganon install via conda).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>raptor is required to build databases with the Hierarchical Interleaved Bloom Filter (<code>ganon build --filter-type hibf</code>)
To build old style ganon indices <code>ganon build --filter-type ibf</code>, raptor is not required</p>
</div>
<p>To install raptor from source, follow the instructions below:</p>
<h4 id="dependencies">Dependencies<a class="headerlink" href="#dependencies" title="Permanent link">:)</a></h4>
<ul>
<li>CMake &gt;= 3.18</li>
<li>GCC 11, 12 or 13 (most recent minor version)</li>
</ul>
<h4 id="downloading-and-building-raptor-submodules">Downloading and building raptor + submodules<a class="headerlink" href="#downloading-and-building-raptor-submodules" title="Permanent link">:)</a></h4>
<pre><code class="language-bash">git clone --branch raptor-v3.0.1 --recurse-submodules https://github.com/seqan/raptor
</code></pre>
<pre><code class="language-bash">cd raptor
mkdir -p build
cd build
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS=&quot;-std=c++23 -Wno-interference-size&quot; ..
make -j 4
</code></pre>
<ul>
<li>binaries will be located in the <code>bin</code> directory</li>
<li>you may have to inform <code>ganon build</code> the path to the binaries with <code>--raptor-path raptor/build/bin</code></li>
</ul>
<h3 id="testing">Testing<a class="headerlink" href="#testing" title="Permanent link">:)</a></h3>
<p>If everything was properly installed, the following command should show the help pages without errors:</p>
<pre><code class="language-bash">ganon -h
</code></pre>
<h4 id="running-tests">Running tests<a class="headerlink" href="#running-tests" title="Permanent link">:)</a></h4>
<pre><code class="language-bash">python3 -m unittest discover -s tests/ganon/integration/
python3 -m unittest discover -s tests/ganon/integration_online/  # optional - downloads large files
cd build/
ctest -VV .
</code></pre>
<h2 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">:)</a></h2>
<pre><code>usage: ganon [-h] [-v]
             {build,build-custom,update,classify,reassign,report,table} ...

- - - - - - - - - -
   _  _  _  _  _   
  (_|(_|| |(_)| |  
   _|   v. 2.0.0
- - - - - - - - - -

positional arguments:
  {build,build-custom,update,classify,reassign,report,table}
    build               Download and build ganon default databases
                        (refseq/genbank)
    build-custom        Build custom ganon databases
    update              Update ganon default databases
    classify            Classify reads against built databases
    reassign            Reassign reads with multiple matches with an EM
                        algorithm
    report              Generate reports from classification results
    table               Generate table from reports

options:
  -h, --help            show this help message and exit
  -v, --version         Show program's version number and exit.
</code></pre>
<details>
  <summary>ganon build</summary>


<pre><code>usage: ganon build [-h] [-g [...]] [-a [...]] [-l] [-b [...]] [-o] [-c] [-r] [-u] [-m [...]] [-z [...]]
                   [--skip-genome-size] -d DB_PREFIX [-x] [-t] [-p] [-k] [-w] [-s] [-f] [-j] [-y] [-v] [--restart]
                   [--verbose] [--quiet] [--write-info-file]

options:
  -h, --help            show this help message and exit

required arguments:
  -g [ ...], --organism-group [ ...]
                        One or more organism groups to download [archaea, bacteria, fungi, human, invertebrate,
                        metagenomes, other, plant, protozoa, vertebrate_mammalian, vertebrate_other, viral]. Mutually
                        exclusive --taxid (default: None)
  -a [ ...], --taxid [ ...]
                        One or more taxonomic identifiers to download. e.g. 562 (-x ncbi) or 's__Escherichia coli' (-x
                        gtdb). Mutually exclusive --organism-group (default: None)
  -d DB_PREFIX, --db-prefix DB_PREFIX
                        Database output prefix (default: None)

database arguments:
  -l , --level          Highest level to build the database. Options: any available taxonomic rank [species, genus,
                        ...], 'leaves' for taxonomic leaves or 'assembly' for a assembly/strain based analysis (default:
                        species)

download arguments:
  -b [ ...], --source [ ...]
                        Source to download [refseq, genbank] (default: ['refseq'])
  -o , --top            Download limited assemblies for each taxa. 0 for all. (default: 0)
  -c, --complete-genomes
                        Download only sub-set of complete genomes (default: False)
  -r, --representative-genomes
                        Download only sub-set of representative genomes (default: False)
  -u , --genome-updater 
                        Additional genome_updater parameters (https://github.com/pirovc/genome_updater) (default: None)
  -m [ ...], --taxonomy-files [ ...]
                        Specific files for taxonomy - otherwise files will be downloaded (default: None)
  -z [ ...], --genome-size-files [ ...]
                        Specific files for genome size estimation - otherwise files will be downloaded (default: None)
  --skip-genome-size    Do not attempt to get genome sizes. Activate this option when using sequences not representing
                        full genomes. (default: False)

important arguments:
  -x , --taxonomy       Set taxonomy to enable taxonomic classification, lca and reports [ncbi, gtdb, skip] (default:
                        ncbi)
  -t , --threads 

advanced arguments:
  -p , --max-fp         Max. false positive for bloom filters. Mutually exclusive --filter-size. Defaults to 0.001 with
                        --filter-type hibf or 0.05 with --filter-type ibf. (default: None)
  -k , --kmer-size      The k-mer size to split sequences. (default: 19)
  -w , --window-size    The window-size to build filter with minimizers. (default: 31)
  -s , --hash-functions 
                        The number of hash functions for the interleaved bloom filter [0-5]. 0 to detect optimal value.
                        (default: 4)
  -f , --filter-size    Fixed size for filter in Megabytes (MB). Mutually exclusive --max-fp. Only valid for --filter-
                        type ibf. (default: 0)
  -j , --mode           Create smaller or faster filters at the cost of classification speed or database size,
                        respectively [avg, smaller, smallest, faster, fastest]. If --filter-size is used,
                        smaller/smallest refers to the false positive rate. By default, an average value is calculated
                        to balance classification speed and database size. Only valid for --filter-type ibf. (default:
                        avg)
  -y , --min-length     Skip sequences smaller then value defined. 0 to not skip any sequence. Only valid for --filter-
                        type ibf. (default: 0)
  -v , --filter-type    Variant of bloom filter to use [hibf, ibf]. hibf requires raptor &gt;= v3.0.1 installed or binary
                        path set with --raptor-path. --mode, --filter-size and --min-length will be ignored with hibf.
                        hibf will set --max-fp 0.001 as default. (default: hibf)

optional arguments:
  --restart             Restart build/update from scratch, do not try to resume from the latest possible step.
                        {db_prefix}_files/ will be deleted if present. (default: False)
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)
  --write-info-file     Save copy of target info generated to {db_prefix}.info.tsv. Can be re-used as --input-file for
                        further attempts. (default: False)
</code></pre>


</details>

<details>
  <summary>ganon build-custom</summary>


<pre><code>usage: ganon build-custom [-h] [-i [...]] [-e] [-c] [-n] [-a] [-l] [-m [...]] [-z [...]] [--skip-genome-size] [-r [...]]
                          [-q [...]] -d DB_PREFIX [-x] [-t] [-p] [-k] [-w] [-s] [-f] [-j] [-y] [-v] [--restart]
                          [--verbose] [--quiet] [--write-info-file]

options:
  -h, --help            show this help message and exit

required arguments:
  -i [ ...], --input [ ...]
                        Input file(s) and/or folder(s). Mutually exclusive --input-file. (default: None)
  -e , --input-extension 
                        Required if --input contains folder(s). Wildcards/Shell Expansions not supported (e.g. *).
                        (default: fna.gz)
  -c, --input-recursive
                        Look for files recursively in folder(s) provided with --input (default: False)
  -d DB_PREFIX, --db-prefix DB_PREFIX
                        Database output prefix (default: None)

custom arguments:
  -n , --input-file     Manually set information for input files: file &lt;tab&gt; [target &lt;tab&gt; node &lt;tab&gt; specialization
                        &lt;tab&gt; specialization name]. target is the sequence identifier if --input-target sequence (file
                        can be repeated for multiple sequences). if --input-target file and target is not set, filename
                        is used. node is the taxonomic identifier. Mutually exclusive --input (default: None)
  -a , --input-target   Target to use [file, sequence]. By default: 'file' if multiple input files are provided or
                        --input-file is set, 'sequence' if a single file is provided. Using 'file' is recommended and
                        will speed-up the building process (default: None)
  -l , --level          Use a specialized target to build the database. By default, --level is the --input-target.
                        Options: any available taxonomic rank [species, genus, ...] or 'leaves' (requires --taxonomy).
                        Further specialization options [assembly, custom]. assembly will retrieve and use the assembly
                        accession and name. custom requires and uses the specialization field in the --input-file.
                        (default: None)
  -m [ ...], --taxonomy-files [ ...]
                        Specific files for taxonomy - otherwise files will be downloaded (default: None)
  -z [ ...], --genome-size-files [ ...]
                        Specific files for genome size estimation - otherwise files will be downloaded (default: None)
  --skip-genome-size    Do not attempt to get genome sizes. Activate this option when using sequences not representing
                        full genomes. (default: False)

ncbi arguments:
  -r [ ...], --ncbi-sequence-info [ ...]
                        Uses NCBI e-utils webservices or downloads accession2taxid files to extract target information.
                        [eutils, nucl_gb, nucl_wgs, nucl_est, nucl_gss, pdb, prot, dead_nucl, dead_wgs, dead_prot or one
                        or more accession2taxid files from https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/].
                        By default uses e-utils up-to 50000 sequences or downloads nucl_gb nucl_wgs otherwise. (default:
                        [])
  -q [ ...], --ncbi-file-info [ ...]
                        Downloads assembly_summary files to extract target information. [refseq, genbank,
                        refseq_historical, genbank_historical or one or more assembly_summary files from
                        https://ftp.ncbi.nlm.nih.gov/genomes/] (default: ['refseq', 'genbank'])

important arguments:
  -x , --taxonomy       Set taxonomy to enable taxonomic classification, lca and reports [ncbi, gtdb, skip] (default:
                        ncbi)
  -t , --threads 

advanced arguments:
  -p , --max-fp         Max. false positive for bloom filters. Mutually exclusive --filter-size. Defaults to 0.001 with
                        --filter-type hibf or 0.05 with --filter-type ibf. (default: None)
  -k , --kmer-size      The k-mer size to split sequences. (default: 19)
  -w , --window-size    The window-size to build filter with minimizers. (default: 31)
  -s , --hash-functions 
                        The number of hash functions for the interleaved bloom filter [0-5]. 0 to detect optimal value.
                        (default: 4)
  -f , --filter-size    Fixed size for filter in Megabytes (MB). Mutually exclusive --max-fp. Only valid for --filter-
                        type ibf. (default: 0)
  -j , --mode           Create smaller or faster filters at the cost of classification speed or database size,
                        respectively [avg, smaller, smallest, faster, fastest]. If --filter-size is used,
                        smaller/smallest refers to the false positive rate. By default, an average value is calculated
                        to balance classification speed and database size. Only valid for --filter-type ibf. (default:
                        avg)
  -y , --min-length     Skip sequences smaller then value defined. 0 to not skip any sequence. Only valid for --filter-
                        type ibf. (default: 0)
  -v , --filter-type    Variant of bloom filter to use [hibf, ibf]. hibf requires raptor &gt;= v3.0.1 installed or binary
                        path set with --raptor-path. --mode, --filter-size and --min-length will be ignored with hibf.
                        hibf will set --max-fp 0.001 as default. (default: hibf)

optional arguments:
  --restart             Restart build/update from scratch, do not try to resume from the latest possible step.
                        {db_prefix}_files/ will be deleted if present. (default: False)
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)
  --write-info-file     Save copy of target info generated to {db_prefix}.info.tsv. Can be re-used as --input-file for
                        further attempts. (default: False)
</code></pre>


</details>

<details>
  <summary>ganon update</summary>


<pre><code>usage: ganon update [-h] -d DB_PREFIX [-o] [-t] [--restart] [--verbose] [--quiet] [--write-info-file]

options:
  -h, --help            show this help message and exit

required arguments:
  -d DB_PREFIX, --db-prefix DB_PREFIX
                        Existing database input prefix (default: None)

important arguments:
  -o , --output-db-prefix 
                        Output database prefix. By default will be the same as --db-prefix and overwrite files (default:
                        None)
  -t , --threads 

optional arguments:
  --restart             Restart build/update from scratch, do not try to resume from the latest possible step.
                        {db_prefix}_files/ will be deleted if present. (default: False)
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)
  --write-info-file     Save copy of target info generated to {db_prefix}.info.tsv. Can be re-used as --input-file for
                        further attempts. (default: False)
</code></pre>


</details>

<details>
  <summary>ganon classify</summary>


<pre><code>usage: ganon classify [-h] -d [DB_PREFIX ...] [-s [reads.fq[.gz] ...]] [-p [reads.1.fq[.gz] reads.2.fq[.gz] ...]]
                      [-c [...]] [-e [...]] [-m] [--ranks [...]] [--min-count] [--report-type] [--skip-report] [-o]
                      [--output-one] [--output-all] [--output-unclassified] [--output-single] [-t] [-b] [-f [...]]
                      [-l [...]] [--verbose] [--quiet]

options:
  -h, --help            show this help message and exit

required arguments:
  -d [DB_PREFIX ...], --db-prefix [DB_PREFIX ...]
                        Database input prefix[es] (default: None)
  -s [reads.fq[.gz] ...], --single-reads [reads.fq[.gz] ...]
                        Multi-fastq[.gz] file[s] to classify (default: None)
  -p [reads.1.fq[.gz] reads.2.fq[.gz] ...], --paired-reads [reads.1.fq[.gz] reads.2.fq[.gz] ...]
                        Multi-fastq[.gz] pairs of file[s] to classify (default: None)

cutoff/filter arguments:
  -c [ ...], --rel-cutoff [ ...]
                        Min. percentage of a read (set of k-mers) shared with a reference necessary to consider a match.
                        Generally used to remove low similarity matches. Single value or one per database (e.g. 0.7 1
                        0.25). 0 for no cutoff (default: [0.75])
  -e [ ...], --rel-filter [ ...]
                        Additional relative percentage of matches (relative to the best match) to keep. Generally used
                        to keep top matches above cutoff. Single value or one per hierarchy (e.g. 0.1 0). 1 for no
                        filter (default: [0.1])

post-processing/report arguments:
  -m , --multiple-matches 
                        Method to solve reads with multiple matches [em, lca, skip]. em -&gt; expectation maximization
                        algorithm based on unique matches. lca -&gt; lowest common ancestor based on taxonomy. The EM
                        algorithm can be executed later with 'ganon reassign' using the .all file (--output-all).
                        (default: em)
  --ranks [ ...]        Ranks to report taxonomic abundances (.tre). empty will report default ranks [superkingdom,
                        phylum, class, order, family, genus, species, assembly]. (default: [])
  --min-count           Minimum percentage/counts to report an taxa (.tre) [use values between 0-1 for percentage, &gt;1
                        for counts] (default: 5e-05)
  --report-type         Type of report (.tre) [abundance, reads, matches, dist, corr]. More info in 'ganon report'.
                        (default: abundance)
  --skip-report         Disable tree-like report (.tre) at the end of classification. Can be done later with 'ganon
                        report'. (default: False)

output arguments:
  -o , --output-prefix 
                        Output prefix for output (.rep) and tree-like report (.tre). Empty to output to STDOUT (only
                        .rep) (default: None)
  --output-one          Output a file with one match for each read (.one) either an unique match or a result from the EM
                        or a LCA algorithm (--multiple-matches) (default: False)
  --output-all          Output a file with all unique and multiple matches (.all) (default: False)
  --output-unclassified
                        Output a file with unclassified read headers (.unc) (default: False)
  --output-single       When using multiple hierarchical levels, output everything in one file instead of one per
                        hierarchy (default: False)

other arguments:
  -t , --threads        Number of sub-processes/threads to use (default: 1)
  -b, --binning         Optimized parameters for binning (--rel-cutoff 0.25 --rel-filter 0 --min-count 0 --report-type
                        reads). Will report sequence abundances (.tre) instead of tax. abundance. (default: False)
  -f [ ...], --fpr-query [ ...]
                        Max. false positive of a query to accept a match. Applied after --rel-cutoff and --rel-filter.
                        Generally used to remove false positives matches querying a database build with large --max-fp.
                        Single value or one per hierarchy (e.g. 0.1 0). 1 for no filter (default: [1e-05])
  -l [ ...], --hierarchy-labels [ ...]
                        Hierarchy definition of --db-prefix files to be classified. Can also be a string, but input will
                        be sorted to define order (e.g. 1 1 2 3). The default value reported without hierarchy is 'H1'
                        (default: None)
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)
</code></pre>


</details>

<details>
  <summary>ganon reassign</summary>


<pre><code>usage: ganon reassign [-h] -i  -o OUTPUT_PREFIX [-e] [-s] [--remove-all] [--skip-one] [--verbose] [--quiet]

options:
  -h, --help            show this help message and exit

required arguments:
  -i , --input-prefix   Input prefix to find files from ganon classify (.all and optionally .rep) (default: None)
  -o OUTPUT_PREFIX, --output-prefix OUTPUT_PREFIX
                        Output prefix for reassigned file (.one and optionally .rep). In case of multiple files, the
                        base input filename will be appended at the end of the output file 'output_prefix +
                        FILENAME.out' (default: None)

EM arguments:
  -e , --max-iter       Max. number of iterations for the EM algorithm. If 0, will run until convergence (check
                        --threshold) (default: 10)
  -s , --threshold      Convergence threshold limit to stop the EM algorithm. (default: 0)

other arguments:
  --remove-all          Remove input file (.all) after processing. (default: False)
  --skip-one            Do not write output file (.one) after processing. (default: False)
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)
</code></pre>


</details>

<details>
  <summary>ganon report</summary>


<pre><code>usage: ganon report [-h] -i [...] [-e INPUT_EXTENSION] -o OUTPUT_PREFIX [-d [...]] [-x] [-m [...]] [-z [...]]
                    [--skip-genome-size] [-f] [-t] [-r [...]] [-s] [-a] [-y] [-p [...]] [-k [...]] [-c] [--verbose]
                    [--quiet] [--min-count] [--max-count] [--names [...]] [--names-with [...]] [--taxids [...]]

options:
  -h, --help            show this help message and exit

required arguments:
  -i [ ...], --input [ ...]
                        Input file(s) and/or folder(s). '.rep' file(s) from ganon classify. (default: None)
  -e INPUT_EXTENSION, --input-extension INPUT_EXTENSION
                        Required if --input contains folder(s). Wildcards/Shell Expansions not supported (e.g. *).
                        (default: rep)
  -o OUTPUT_PREFIX, --output-prefix OUTPUT_PREFIX
                        Output prefix for report file 'output_prefix.tre'. In case of multiple files, the base input
                        filename will be appended at the end of the output file 'output_prefix + FILENAME.tre' (default:
                        None)

db/tax arguments:
  -d [ ...], --db-prefix [ ...]
                        Database prefix(es) used for classification. Only '.tax' file(s) are required. If not provided,
                        new taxonomy will be downloaded. Mutually exclusive with --taxonomy. (default: [])
  -x , --taxonomy       Taxonomy database to use [ncbi, gtdb, skip]. Mutually exclusive with --db-prefix. (default:
                        ncbi)
  -m [ ...], --taxonomy-files [ ...]
                        Specific files for taxonomy - otherwise files will be downloaded (default: None)
  -z [ ...], --genome-size-files [ ...]
                        Specific files for genome size estimation - otherwise files will be downloaded (default: None)
  --skip-genome-size    Do not attempt to get genome sizes. Valid only without --db-prefix. Activate this option when
                        using sequences not representing full genomes. (default: False)

output arguments:
  -f , --output-format 
                        Output format [text, tsv, csv, bioboxes]. text outputs a tabulated formatted text file for
                        better visualization. bioboxes is the the CAMI challenge profiling format (only
                        percentage/abundances are reported). (default: tsv)
  -t , --report-type    Type of report [abundance, reads, matches, dist, corr]. 'abundance' -&gt; tax. abundance (re-
                        distribute read counts and correct by genome size), 'reads' -&gt; sequence abundance, 'matches' -&gt;
                        report all unique and shared matches, 'dist' -&gt; like reads with re-distribution of shared read
                        counts only, 'corr' -&gt; like abundance without re-distribution of shared read counts (default:
                        abundance)
  -r [ ...], --ranks [ ...]
                        Ranks to report ['', 'all', custom list]. 'all' for all possible ranks. empty for default ranks
                        [superkingdom, phylum, class, order, family, genus, species, assembly]. (default: [])
  -s , --sort           Sort report by [rank, lineage, count, unique]. Default: rank (with custom --ranks) or lineage
                        (with --ranks all) (default: )
  -a, --no-orphan       Omit orphan nodes from the final report. Otherwise, orphan nodes (= nodes not found in the
                        db/tax) are reported as 'na' with root as direct parent. (default: False)
  -y, --split-hierarchy
                        Split output reports by hierarchy (from ganon classify --hierarchy-labels). If activated, the
                        output files will be named as '{output_prefix}.{hierarchy}.tre' (default: False)
  -p [ ...], --skip-hierarchy [ ...]
                        One or more hierarchies to skip in the report (from ganon classify --hierarchy-labels) (default:
                        [])
  -k [ ...], --keep-hierarchy [ ...]
                        One or more hierarchies to keep in the report (from ganon classify --hierarchy-labels) (default:
                        [])
  -c , --top-percentile 
                        Top percentile filter, based on percentage/relative abundance. Applied only at default ranks
                        [superkingdom, phylum, class, order, family, genus, species, assembly] (default: 0)

optional arguments:
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)

filter arguments:
  --min-count           Minimum number/percentage of counts to keep an taxa [values between 0-1 for percentage, &gt;1
                        specific number] (default: 0)
  --max-count           Maximum number/percentage of counts to keep an taxa [values between 0-1 for percentage, &gt;1
                        specific number] (default: 0)
  --names [ ...]        Show only entries matching exact names of the provided list (default: [])
  --names-with [ ...]   Show entries containing full or partial names of the provided list (default: [])
  --taxids [ ...]       One or more taxids to report (including children taxa) (default: [])
</code></pre>


</details>

<details>
  <summary>ganon table</summary>


<pre><code>usage: ganon table [-h] -i [...] [-e] -o OUTPUT_FILE [-l] [-f] [-t] [-a] [-m] [-r] [-n] [--header]
                   [--unclassified-label] [--filtered-label] [--skip-zeros] [--transpose] [--verbose] [--quiet]
                   [--min-count] [--max-count] [--names [...]] [--names-with [...]] [--taxids [...]]

options:
  -h, --help            show this help message and exit

required arguments:
  -i [ ...], --input [ ...]
                        Input file(s) and/or folder(s). '.tre' file(s) from ganon report. (default: None)
  -e , --input-extension 
                        Required if --input contains folder(s). Wildcards/Shell Expansions not supported (e.g. *).
                        (default: tre)
  -o OUTPUT_FILE, --output-file OUTPUT_FILE
                        Output filename for the table (default: None)

output arguments:
  -l , --output-value   Output value on the table [percentage, counts]. percentage values are reported between [0-1]
                        (default: counts)
  -f , --output-format 
                        Output format [tsv, csv] (default: tsv)
  -t , --top-sample     Top hits of each sample individually (default: 0)
  -a , --top-all        Top hits of all samples (ranked by percentage) (default: 0)
  -m , --min-frequency 
                        Minimum number/percentage of files containing an taxa to keep the taxa [values between 0-1 for
                        percentage, &gt;1 specific number] (default: 0)
  -r , --rank           Define specific rank to report. Empty will report all ranks. (default: None)
  -n, --no-root         Do not report root node entry and lineage. Direct and shared matches to root will be accounted
                        as unclassified (default: False)
  --header              Header information [name, taxid, lineage] (default: name)
  --unclassified-label 
                        Add column with unclassified count/percentage with the chosen label. May be the same as
                        --filtered-label (e.g. unassigned) (default: None)
  --filtered-label      Add column with filtered count/percentage with the chosen label. May be the same as
                        --unclassified-label (e.g. unassigned) (default: None)
  --skip-zeros          Do not print lines with only zero count/percentage (default: False)
  --transpose           Transpose output table (taxa as cols and files as rows) (default: False)

optional arguments:
  --verbose             Verbose output mode (default: False)
  --quiet               Quiet output mode (default: False)

filter arguments:
  --min-count           Minimum number/percentage of counts to keep an taxa [values between 0-1 for percentage, &gt;1
                        specific number] (default: 0)
  --max-count           Maximum number/percentage of counts to keep an taxa [values between 0-1 for percentage, &gt;1
                        specific number] (default: 0)
  --names [ ...]        Show only entries matching exact names of the provided list (default: [])
  --names-with [ ...]   Show entries containing full or partial names of the provided list (default: [])
  --taxids [ ...]       One or more taxids to report (including children taxa) (default: [])
</code></pre>


</details>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="start/" class="btn btn-neutral float-right" title="Quick Start">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
      <span><a href="start/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>

<!--
MkDocs version : 1.5.1
Build Date UTC : 2023-11-17 16:39:25.603777+00:00
-->
