<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Tutorials - ganon</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Tutorials";
        var mkdocs_page_input_path = "tutorials.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> ganon
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">ganon</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../default_databases/">Default databases</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../custom_databases/">Custom databases</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../classification/">Classification</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../reports/">Reports</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../outputfiles/">Output files</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Tutorials</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#custom-build">Custom build</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#choosing-and-explaining-parameters">Choosing and explaining parameters</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#ganon-build">ganon build</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#filter-false-positive-and-size-max-fp-filter-size">filter false positive and size (--max-fp, --filter-size)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#minimizers-window-size-kmer-size">minimizers (--window-size, --kmer-size)</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#ganon-classify">ganon classify</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#reads-single-reads-paired-reads">reads (--single-reads, --paired-reads)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cutoff-and-filter-rel-cutoff-rel-filter">cutoff and filter (--rel-cutoff, --rel-filter)</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#ganon-build-custom">ganon build-custom</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#target-file-or-sequence-input-target">Target file or sequence (--input-target)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#build-level-level">Build level (--level)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#genome-sizes-genome-size-files">Genome sizes (--genome-size-files)</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#retrieving-info-ncbi-sequence-info-ncbi-file-info">Retrieving info (--ncbi-sequence-info, --ncbi-file-info)</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#ganon-report">ganon report</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#report-type-report-type">report type (--report-type)</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#commonly-used-reference-set-for-metagenomics-analysis-archaeabacteriafungiviral-from-ncbi-refseq-complete-genomes-one-assembly-per-taxa">Commonly used reference set for metagenomics analysis (archaea+bacteria+fungi+viral from NCBI RefSeq, complete genomes, one assembly per taxa)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#gtdb-database-one-assembly-per-taxa">GTDB database, one assembly per taxa</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#database-based-on-specific-taxonomic-identifiers-203492-fusobacteriaceae">Database based on specific taxonomic identifiers (203492 - Fusobacteriaceae)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#customized-database-at-assembly-level">Customized database at assembly level</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#customized-database-at-species-level-build-based-on-files-previously-downloaded-with-genome_updater">Customized database at species level build based on files previously downloaded with genome_updater</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#customized-database-with-sequence-as-target-to-classify-reads-at-sequence-level">Customized database with sequence as target (to classify reads at sequence level)</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">ganon</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>Tutorials</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="tutorials">Tutorials</h1>
<h2 id="custom-build">Custom build</h2>
<h2 id="choosing-and-explaining-parameters">Choosing and explaining parameters</h2>
<p>The most important parameters and trade-offs are:</p>
<ul>
<li><code>ganon build</code> <code>--window-size --kmer-size</code>: the <em>window</em> value should always be the same or larger than the <em>kmer</em> value. The larger the difference between them, the smaller the database will be. However, some sensitivity/precision loss in classification is expected with small <em>kmer</em> and/or large <em>window</em>. Larger <em>kmer</em> values (e.g. <code>31</code>) will improve classification, specially read binning, at a cost of way bigger databases.</li>
<li><code>ganon classify</code> <code>--rel-cutoff</code>: this value defines the threshold for matches between reads and database. Higher <code>--rel-cutoff</code> values will improve precision and decrease sensitivity with expected less unique matches but an increase in overall matches. For taxonomic profiling, a higher value between <code>0.4</code> and <code>0.8</code> may provide better results. For read binning, lower values between <code>0.2</code> and <code>0.4</code> are recommended.</li>
<li><code>ganon classify</code> <code>--rel-filter</code>: further filter top matches after cutoff is applied. Usually set between <code>0</code> and <code>0.2</code>.</li>
<li><code>ganon classify</code> <code>--reassign</code>: runs an EM-algorithm to reassign reads that received multiple matches. It provides a unique match for each read at the level the database was built (e.g. assembly or species). Mostly useful for read binning, with little overall impact on taxonomic profiling. Can be used independently with <code>ganon reassign</code>.</li>
<li><code>ganon report</code> <code>--report-type</code>: reports either taxonomic, sequence or matches abundances. Use <code>corr</code> or <code>abundance</code> for taxonomic profiling, <code>reads</code> or <code>dist</code> for sequence profiling and <code>matches</code> to report a summary of all matches.</li>
<li><code>ganon report</code> <code>--min-count</code>: cutoff to discard underrepresented taxa. Useful to remove the common long tail of spurious matches and false positives when performing classification. Values between <code>0.0001</code> (0.01%) and <code>0.001</code> (0.1%) improved sensitivity and precision in our evaluations. The higher the value, the more precise the outcome, with a sensitivity loss. Alternatively <code>--top-percentile</code> can be used to keep a relative amount of taxa instead a hard cutoff.</li>
</ul>
<p>The numeric values above are averages from several experiments with different sample types and database contents. They may not work as expected for your data. If you are not sure which values to use or see something unexpected, please open an <a href="https://github.com/pirovc/ganon/issues">issue</a>.</p>
<p>Below, a more in-depth explanation of important parameters:</p>
<h3 id="ganon-build">ganon build</h3>
<h4 id="filter-false-positive-and-size-max-fp-filter-size">filter false positive and size (--max-fp, --filter-size)</h4>
<p>ganon indices are based on bloom filters and can have false positive matches. This can be controlled with <code>--max-fp</code> parameter. The lower the <code>--max-fp</code>, the less chances of false positives, but the larger the database size will be. For example, with <code>--max-fp 0.01</code> the database will be build so any target (e.g. assembly, specificed with <code>--level</code>) will have 1 in a 100 change of reporting a false match (between minimizer k-mers).</p>
<p>Alternatively, one can set a specific size for the final index with <code>--filer-size</code>. When using this option, please observe the theoretic false positive of the index reported at the end of the building process.</p>
<h4 id="minimizers-window-size-kmer-size">minimizers (--window-size, --kmer-size)</h4>
<p>in <code>ganon build</code>, when <code>--window-size</code> &gt; <code>--kmer-size</code> minimizers are used. That means that for a every window, a single k-mer will be selected. It produces smaller database files and requires substantially less memory overall. It may increase building times but will have a huge benefit for classification times. Sensitivity and precision can be reduced by small margins. If <code>--window-size</code> = <code>--kmer-size</code>, all k-mers are going to be used to build the database.</p>
<h3 id="ganon-classify">ganon classify</h3>
<h4 id="reads-single-reads-paired-reads">reads (--single-reads, --paired-reads)</h4>
<p>ganon accepts single-end and paired-end reads. Both types can be use at the same time. In paired-end mode, reads are always reported with the header of the first pair. Paired-end reads are classified in a standard forward-reverse orientation. The max. read length accepted is 65535 (accounting both reads in paired mode).</p>
<h4 id="cutoff-and-filter-rel-cutoff-rel-filter">cutoff and filter (--rel-cutoff, --rel-filter)</h4>
<p>ganon has two parameters to control a match between reads and references: <code>--rel-cutoff</code> and <code>--rel-filter</code>.</p>
<p>Every read can be classified against none, one or more references. What will be reported is the remaining matches after <code>cutoff</code> and <code>filter</code> thresholds are applied, based on the number of shared minimizers (or k-mers) between sequences.</p>
<p>The <code>cutoff</code> is the first. It should be set as a minimal value to consider a match between a read and a reference. Next the <code>filter</code> is applied to the remaining matches. <code>filter</code> thresholds are relative to the best scoring match and control how far from the best match further matches are allowed. <code>cutoff</code> can be interpreted as the lower bound to discard spurious matches and <code>filter</code> as the fine tuning to control what to keep.</p>
<p>For example, using <code>--kmer-size 19</code> (and <code>--window-size 19</code> to simplify the example), a certain read (100bp) has the following matches with the 5 references (<code>ref1..5</code>), sorted by shared k-mers:</p>
<table>
<thead>
<tr>
<th>reference</th>
<th>shared k-mers</th>
</tr>
</thead>
<tbody>
<tr>
<td>ref1</td>
<td>82</td>
</tr>
<tr>
<td>ref2</td>
<td>68</td>
</tr>
<tr>
<td>ref3</td>
<td>44</td>
</tr>
<tr>
<td>ref4</td>
<td>25</td>
</tr>
<tr>
<td>ref5</td>
<td>20</td>
</tr>
</tbody>
</table>
<p>this read can have at most 82 shared k-mers (<code>100-19+1=82</code>). With <code>--rel-cutoff 0.25</code>, the following matches will be discarded:</p>
<table>
<thead>
<tr>
<th>reference</th>
<th>shared k-mers</th>
<th>--rel-cutoff 0.25</th>
</tr>
</thead>
<tbody>
<tr>
<td>ref1</td>
<td>82</td>
<td></td>
</tr>
<tr>
<td>ref2</td>
<td>68</td>
<td></td>
</tr>
<tr>
<td>ref3</td>
<td>44</td>
<td></td>
</tr>
<tr>
<td>ref4</td>
<td>25</td>
<td></td>
</tr>
<tr>
<td>~~ref5~~</td>
<td>~~20~~</td>
<td>X</td>
</tr>
</tbody>
</table>
<p>since the <code>--rel-cutoff</code> threshold is <code>82 * 0.25 = 21</code> (ceiling is applied). Further, with <code>--rel-filter 0.3</code>, the following matches will be discarded:</p>
<table>
<thead>
<tr>
<th>reference</th>
<th>shared k-mers</th>
<th>--rel-cutoff 0.25</th>
<th>--rel-filter 0.3</th>
</tr>
</thead>
<tbody>
<tr>
<td>ref1</td>
<td>82</td>
<td></td>
<td></td>
</tr>
<tr>
<td>ref2</td>
<td>68</td>
<td></td>
<td></td>
</tr>
<tr>
<td>~~ref3~~</td>
<td>~~44~~</td>
<td></td>
<td>X</td>
</tr>
<tr>
<td>~~ref4~~</td>
<td>~~25~~</td>
<td></td>
<td>X</td>
</tr>
<tr>
<td>~~ref5~~</td>
<td>~~20~~</td>
<td>X</td>
<td></td>
</tr>
</tbody>
</table>
<p>since best match is 82, the filter parameter is removing any match below <code>0.3 * 82 = 57</code> (ceiling is applied) shared k-mers. <code>ref1</code> and <code>ref2</code> are reported as matches.</p>
<p>For databases built with <code>--window-size</code>, the relative values are not based on the maximum number of possible shared k-mers but on the actual number of unique minimizers extracted from the read.</p>
<p>A different <code>cutoff</code> can be set for every database in a multiple or hierarchical database classification. A different <code>filter</code> can be set for every level of a hierarchical database classification.</p>
<p>Note that reads that remain with only one reference match (after <code>cutoff</code> and <code>filter</code> are applied) are considered a unique match.</p>
<h3 id="ganon-build-custom">ganon build-custom</h3>
<h4 id="target-file-or-sequence-input-target">Target file or sequence (--input-target)</h4>
<p>Customized builds can be done either by file or sequence. <code>--input-target file</code> will consider every file provided with <code>--input</code> a single unit. <code>--input-target sequence</code> will use every sequence as a unit.</p>
<p><code>--input-target file</code> is the default behavior and most efficient way to build databases. <code>--input-target sequence</code> should only be used when the input sequences are stored in a single file or when classification at sequence level is desired.</p>
<h4 id="build-level-level">Build level (--level)</h4>
<p>The <code>--level</code> parameter defines the max. depth of the database for classification. This parameter is relevant because the <code>--max-fp</code> is going to be guaranteed at the <code>--level</code> chosen. By default, the level will be the same as <code>--input-target</code>, meaning that classification will be done either at file or sequence level.</p>
<p>Alternatively, <code>--level assembly</code> will link the file or sequence target information with assembly accessions retrieved from NCBI servers. <code>--level leaves</code> or <code>--level species</code> (or genus, family, ...) will link the targets with taxonomic information and prune the tree at the chosen level. <code>--level custom</code> will use specialization level define in the <code>--input-file</code>.</p>
<h4 id="genome-sizes-genome-size-files">Genome sizes (--genome-size-files)</h4>
<p>Ganon will automatically download auxiliary files to define an approximate genome size for each entry in the taxonomic tree. For <code>--taxonomy ncbi</code> the <a href="https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/">species_genome_size.txt.gz</a> is used. For <code>--taxonomy gtdb</code> the <a href="https://data.gtdb.ecogenomic.org/releases/latest/">*_metadata.tar.gz</a> files are used. Those files can be directly provided with the <code>--genome-size-files</code> argument.</p>
<p>Genome sizes of parent nodes are calculated as the average of the respective children nodes. Other nodes without direct assigned genome sizes will use the closest parent with a pre-calculated genome size. The genome sizes are stored in the <a href="#buildupdate">ganon database</a>.</p>
<h4 id="retrieving-info-ncbi-sequence-info-ncbi-file-info">Retrieving info (--ncbi-sequence-info, --ncbi-file-info)</h4>
<p>Further taxonomy and assembly linking information has to be collected to properly build the database. <code>--ncbi-sequence-info</code> and <code>--ncbi-file-info</code> allow customizations on this step.</p>
<p>When <code>--input-target sequence</code>, <code>--ncbi-sequence-info</code> argument allows the use of NCBI e-utils webservices (<code>eutils</code>) or downloads accession2taxid files to extract target information (options <code>nucl_gb</code> <code>nucl_wgs</code> <code>nucl_est</code> <code>nucl_gss</code> <code>pdb</code> <code>prot</code> <code>dead_nucl</code> <code>dead_wgs</code> <code>dead_prot</code>). By default, ganon uses <code>eutils</code> up-to 50000 input sequences, otherwise it downloads <code>nucl_gb</code> <code>nucl_wgs</code> from https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/. Previously downloaded files can be directly provided with this argument.</p>
<p>When <code>--input-target file</code>, <code>--ncbi-file-info</code> uses <code>assembly_summary.txt</code> from https://ftp.ncbi.nlm.nih.gov/genomes/ to extract target information (options <code>refseq</code> <code>genbank</code> <code>refseq_historical</code> <code>genbank_historical</code>. Previously downloaded files can be directly provided with this argument.</p>
<p>If you are using outdated, removed or inactive assembly or sequence files and accessions from NCBI, make sure to include <code>dead_nucl</code> <code>dead_wgs</code> for <code>--ncbi-sequence-info</code> or <code>refseq_historical</code> <code>genbank_historical</code> for <code>--ncbi-file-info</code>. <code>eutils</code> option does not work with outdated accessions.</p>
<h3 id="ganon-report">ganon report</h3>
<h4 id="report-type-report-type">report type (--report-type)</h4>
<p>Several reports are availble with <code>--report-type</code>: <code>reads</code>, <code>abundance</code>, <code>dist</code>, <code>corr</code>, <code>matches</code>:</p>
<p><code>reads</code> reports <strong>sequence abundances</strong> which are the basic proportion of reads classified in the sample.</p>
<p><code>abundance</code> will convert sequence abundance into <strong>taxonomic abundances</strong> by re-distributing read counts among leaf nodes and correcting by genome size. The re-distribution applies for reads classified with a LCA assignment and it is proportional to the number of unique matches of leaf nodes available in the ganon database (relative to the LCA node). Genome size is estimated based on <a href="#genome-sizes---genome-size-files">NCBI or GTDB auxiliary files</a>. Genome size correction is applied by rank based on default ranks only (superkingdom phylum class order family genus species assembly). Read counts in intermediate ranks will be corrected based on the closest parent default rank and re-assigned to its original rank.</p>
<p><code>dist</code> is the same of <code>reads</code> with read count re-distribution</p>
<p><code>corr</code> is the same of <code>reads</code> with correction by genome size</p>
<p><code>matches</code> will report the total number of matches classified, either unique or shared. <em>This report will output the total number of matches instead the total number of reads reported in all other reports.</em></p>
<h2 id="examples">Examples</h2>
<h3 id="commonly-used-reference-set-for-metagenomics-analysis-archaeabacteriafungiviral-from-ncbi-refseq-complete-genomes-one-assembly-per-taxa">Commonly used reference set for metagenomics analysis (archaea+bacteria+fungi+viral from NCBI RefSeq, complete genomes, one assembly per taxa)</h3>
<pre><code class="language-bash">ganon build --db-prefix abfv_rs_cg_t1a --organism-group archaea bacteria fungi viral --source refseq --taxonomy ncbi --complete-genomes --threads 12 --top 1
</code></pre>
<h3 id="gtdb-database-one-assembly-per-taxa">GTDB database, one assembly per taxa</h3>
<pre><code class="language-bash">ganon build --db-prefix complete_gtdb --organism-group archaea bacteria --source refseq genbank --taxonomy gtdb --threads 12 --top 1
</code></pre>
<h3 id="database-based-on-specific-taxonomic-identifiers-203492-fusobacteriaceae">Database based on specific taxonomic identifiers (203492 - Fusobacteriaceae)</h3>
<pre><code class="language-bash"># NCBI
ganon build --db-prefix fuso_ncbi --taxid &quot;203492&quot; --source refseq genbank --taxonomy ncbi --threads 12
# GTDB
ganon build --db-prefix fuso_gtdb --taxid &quot;f__Fusobacteriaceae&quot; --source refseq genbank --taxonomy gtdb --threads 12
</code></pre>
<h3 id="customized-database-at-assembly-level">Customized database at assembly level</h3>
<pre><code class="language-bash">ganon build-custom --db-prefix my_db --input my_big_fasta_file.fasta.gz --level assembly --threads 12
</code></pre>
<h3 id="customized-database-at-species-level-build-based-on-files-previously-downloaded-with-genome_updater">Customized database at species level build based on files previously downloaded with <a href="https://github.com/pirovc/genome_updater">genome_updater</a></h3>
<pre><code class="language-bash">ganon build-custom --db-prefix custom_db_gu --input myfiles/2022-06-28_10-02-14/files/ --level species --ncbi-file-info outfolder/2022-06-28_10-02-14/assembly_summary.txt --threads 12
</code></pre>
<h3 id="customized-database-with-sequence-as-target-to-classify-reads-at-sequence-level">Customized database with sequence as target (to classify reads at sequence level)</h3>
<pre><code class="language-bash">ganon build-custom --db-prefix seq_target --input myfiles/2022-06-28_10-02-14/files/ --ncbi-file-info outfolder/2022-06-28_10-02-14/assembly_summary.txt --input-target sequence --threads 12
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../outputfiles/" class="btn btn-neutral float-left" title="Output files"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../outputfiles/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
