<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Custom databases (ganon build-custom) - ganon2</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Custom databases (ganon build-custom)";
        var mkdocs_page_input_path = "custom_databases.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> ganon2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">ganon2</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../start/">Quick Start</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../default_databases/">Databases (ganon build)</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Custom databases (ganon build-custom)</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#non-standard-filesheaders-with-input-file">Non-standard files/headers with --input-file</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#examples-of-input-file-using-the-default-input-target-file">Examples of --input-file using the default --input-target file</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#list-of-files">List of files</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#list-of-files-with-alternative-names">List of files with alternative names</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#files-and-taxonomy">Files and taxonomy</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#files-taxonomy-and-specialization">Files, taxonomy and specialization</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#examples-of-input-file-using-input-target-sequence">Examples of --input-file using --input-target sequence</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#sequences-and-taxonomy">Sequences and taxonomy</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#sequences-taxonomy-and-specialization">Sequences, taxonomy and specialization</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#humgut">HumGut</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#plasmid-plastid-and-mitochondrion-from-refseq">Plasmid, Plastid and Mitochondrion from RefSeq</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#univec-univec_core">UniVec, UniVec_core</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#mgnify-genome-catalogues-mags">MGnify genome catalogues (MAGs)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pathogen-detection-fda-argos">Pathogen detection FDA-ARGOS</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#blast-databases-nt-env_nt-nt_prok">BLAST databases (nt env_nt nt_prok ...)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#files-from-genome_updater">Files from genome_updater</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#parameter-details">Parameter details</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#false-positive-and-size-max-fp-filter-size">False positive and size (--max-fp, --filter-size)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#minimizers-window-size-kmer-size">minimizers (--window-size, --kmer-size)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#target-file-or-sequence-input-target">Target file or sequence (--input-target)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#build-level-level">Build level (--level)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#genome-sizes-genome-size-files">Genome sizes (--genome-size-files)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#retrieving-info-ncbi-sequence-info-ncbi-file-info">Retrieving info (--ncbi-sequence-info, --ncbi-file-info)</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../classification/">Classification (ganon classify)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../reports/">Reports (ganon report)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../table/">Table (ganon table)</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../outputfiles/">Output files</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../tutorials/">Tutorials</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">ganon2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>Custom databases (ganon build-custom)</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="custom-databases">Custom databases<a class="headerlink" href="#custom-databases" title="Permanent link">:)</a></h1>
<p>Besides the automated download and build (<code>ganon build</code>) ganon provides a highly customizable build procedure (<code>ganon build-custom</code>) to create databases from local sequence files. The usage of this procedure depends on the configuration of your files:</p>
<ul>
<li>Filename like <code>GCA_002211645.1_ASM221164v1_genomic.fna.gz</code>: genomic fasta files in the NCBI standard, with assembly accession in the beginning of the filename. Provide the files with the <code>--input</code> parameter. ganon will try to retrieve all necessary information to build the database.</li>
<li>Headers like <code>&gt;NC_006297.1 Bacteroides fragilis YCH46 ...</code>: sequence headers are in the NCBI standard, with sequence accession in after <code>&gt;</code> and with a space afterwards (or line break). Provide the files with the <code>--input</code> parameter and set <code>--input-target sequence</code>. ganon will try to retrieve all necessary information to build the database.</li>
<li>For non-standard filenames and headers, <a href="#non-standard-filesheaders-with-input-file">follow this</a></li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><code>--input-target sequence</code> will be slower to build and will use more disk space, since files have be re-written separately for each sequence. More information about building by file or sequence can be found <a href="#target-file-or-sequence-input-target">here</a>.</p>
</div>
<p>The <code>--level</code> is a important parameter that will define the (max.) classification level for the database (<a href="#build-level-level">more infos</a>):</p>
<ul>
<li><code>--level file</code> or <code>sequence</code> -&gt; default behavior (depending on <code>--input-target</code>), use file/sequence as classification target</li>
<li><code>--level assembly</code> -&gt; will retrieve assembly related to the file/sequence, use assembly as classification target</li>
<li><code>--level leaves</code> or <code>species</code>,<code>genus</code>,... -&gt; group input by taxonomy, use tax. nodes at the rank chosen as classification target</li>
</ul>
<p>More infos about other parameters <a href="#parameter-details">here</a>.</p>
<h2 id="non-standard-filesheaders-with-input-file">Non-standard files/headers with <code>--input-file</code><a class="headerlink" href="#non-standard-filesheaders-with-input-file" title="Permanent link">:)</a></h2>
<p>Alternatively to the automatic input methods, it is possible to manually define the input with either standard or <strong>non-standard filenames, accessions and headers</strong> to build custom databases with <code>--input-file</code>. This file should contain the following fields (tab-separated):</p>
<p><code>file [&lt;tab&gt; target &lt;tab&gt; node &lt;tab&gt; specialization &lt;tab&gt; specialization_name].</code></p>
<ul>
<li><code>file</code>: relative or full path to the sequence file</li>
<li><code>target</code>: any unique text to name the file, to be used in the taxonomy</li>
<li><code>node</code>: taxonomic node (e.g. taxid) to link entry with taxonomy</li>
<li><code>specialization</code>: creates a specialized taxonomic level with a custom name, allowing files to be grouped</li>
<li><code>specialization_name</code>: a name for the specialization, to be used in the taxonomy</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>the <code>target</code> and <code>specialization</code> fields (2nd and 4th col) cannot be the same as the <code>node</code> (3rd col)</p>
</div>
<p>Below you find example of <code>--input-file</code>. Note they are slightly different depending on the <code>--input-target</code> chosen. They need to be <em>tab-separated</em> to be properly parsed (tsv).</p>
<h3 id="examples-of-input-file-using-the-default-input-target-file">Examples of <code>--input-file</code> using the default <code>--input-target file</code><a class="headerlink" href="#examples-of-input-file-using-the-default-input-target-file" title="Permanent link">:)</a></h3>
<h4 id="list-of-files">List of files<a class="headerlink" href="#list-of-files" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta
others.fasta
</code></pre>
<p>No taxonomic information is provided so <code>--taxonomy skip</code> should be set. The classification against the generated database will be performed at file level (<code>--level file</code>), since that is the only available information given.</p>
<h4 id="list-of-files-with-alternative-names">List of files with alternative names<a class="headerlink" href="#list-of-files-with-alternative-names" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta  sequences
others.fasta     others
</code></pre>
<p>Just like above, but with a specific name to be used for each file.</p>
<h4 id="files-and-taxonomy">Files and taxonomy<a class="headerlink" href="#files-and-taxonomy" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta  sequences  562
others.fasta     others     623
</code></pre>
<p>The classification max. level against this database will depend on the value set for <code>--level</code>:</p>
<ul>
<li><code>--level file</code> -&gt; use the file (named with target) with node as parent</li>
<li><code>--level leaves</code> or <code>species</code>,<code>genus</code>,... -&gt; files are grouped by taxonomy</li>
</ul>
<h4 id="files-taxonomy-and-specialization">Files, taxonomy and specialization<a class="headerlink" href="#files-taxonomy-and-specialization" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta  sequences  562  ID44444  Escherichia coli TW10119
others.fasta     others     623  ID55555  Shigella flexneri 1a
</code></pre>
<p>The classification max. level against this database will depend on the value set for <code>--level</code>:</p>
<ul>
<li><code>--level custom</code> -&gt; use the specialization (named with specialization_name) with node as parent</li>
<li><code>--level file</code> -&gt; use the file (named with target) as a tax. node as parent</li>
<li><code>--level leaves</code> or <code>species</code>,<code>genus</code>,... -&gt; files are grouped by taxonomy</li>
</ul>
<h3 id="examples-of-input-file-using-input-target-sequence">Examples of <code>--input-file</code> using <code>--input-target sequence</code><a class="headerlink" href="#examples-of-input-file-using-input-target-sequence" title="Permanent link">:)</a></h3>
<p>To provide a tabular information for every sequence in your files, you need to use the <code>target</code> field (2nd col.) of the <code>--input-file</code> to input sequence headers. For example:</p>
<h4 id="sequences-and-taxonomy">Sequences and taxonomy<a class="headerlink" href="#sequences-and-taxonomy" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta  NZ_CP054001.1  562
sequences.fasta  NZ_CP117955.1  623
others.fasta     header1        666
others.fasta     header2        666
</code></pre>
<p>The classification max. level against this database will depend on the value set for <code>--level</code>:</p>
<ul>
<li><code>--level sequence</code> -&gt; use the sequence header with node as parent</li>
<li><code>--level assembly</code> -&gt; will attempt to retrieve the assembly related to the sequence with node as parent</li>
<li><code>--level leaves</code> or <code>species</code>,<code>genus</code>,... -&gt; files are grouped by taxonomy</li>
</ul>
<h4 id="sequences-taxonomy-and-specialization">Sequences, taxonomy and specialization<a class="headerlink" href="#sequences-taxonomy-and-specialization" title="Permanent link">:)</a></h4>
<pre><code>sequences.fasta  NZ_CP054001.1  562  ID44444  Escherichia coli TW10119
sequences.fasta  NZ_CP117955.1  623  ID55555  Shigella flexneri 1a
others.fasta     header1        666  StrainA  My Strain
others.fasta     header2        666  StrainA  My Strain
</code></pre>
<p>The classification max. level against this database will depend on the value set for <code>--level</code>:</p>
<ul>
<li><code>--level custom</code> -&gt; use the specialization (named with specialization_name) with node as parent</li>
<li><code>--level sequence</code> -&gt; use the sequence header with node as parent</li>
<li><code>--level leaves</code> or <code>species</code>,<code>genus</code>,... -&gt; files are grouped by taxonomy</li>
</ul>
<h2 id="examples">Examples<a class="headerlink" href="#examples" title="Permanent link">:)</a></h2>
<p>Below you will find some examples from commonly used repositories for metagenomics analysis with <code>ganon build-custom</code>: </p>
<h3 id="humgut">HumGut<a class="headerlink" href="#humgut" title="Permanent link">:)</a></h3>
<p>Collection of &gt;30000 genomes from healthy human metagenomes. <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-021-01114-w">Article</a>/<a href="https://arken.nmbu.no/~larssn/humgut/">Website</a>.</p>
<pre><code class="language-bash"># Download sequence files
wget --quiet --show-progress &quot;http://arken.nmbu.no/~larssn/humgut/HumGut.tar.gz&quot;
tar xf HumGut.tar.gz 

# Download taxonomy and metadata files
wget &quot;https://arken.nmbu.no/~larssn/humgut/ncbi_nodes.dmp&quot;
wget &quot;https://arken.nmbu.no/~larssn/humgut/ncbi_names.dmp&quot;
wget &quot;https://arken.nmbu.no/~larssn/humgut/HumGut.tsv&quot;
# Generate --input-file from metadata
tail -n+2 HumGut.tsv | awk -F&quot;\t&quot; '{print &quot;fna/&quot;$21&quot;\t&quot;$1&quot;\t&quot;$6}' &gt; HumGut_ganon_input_file.tsv

# Build ganon database
ganon build-custom --input-file HumGut_ganon_input_file.tsv --taxonomy-files ncbi_nodes.dmp ncbi_names.dmp --db-prefix HumGut --level strain --threads 32
</code></pre>
<p>Similarly using GTDB taxonomy files:</p>
<pre><code class="language-bash"># Download taxonomy files
wget &quot;https://arken.nmbu.no/~larssn/humgut/gtdb_nodes.dmp&quot;
wget &quot;https://arken.nmbu.no/~larssn/humgut/gtdb_names.dmp&quot;

# Build ganon database
ganon build-custom --input-file HumGut_ganon_input_file.tsv --taxonomy-files gtdb_nodes.dmp gtdb_names.dmp --db-prefix HumGut_gtdb --level strain --threads 32
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>There is no need to use ganon's gtdb integration here since GTDB files in NCBI format are available</p>
</div>
<h3 id="plasmid-plastid-and-mitochondrion-from-refseq">Plasmid, Plastid and Mitochondrion from RefSeq<a class="headerlink" href="#plasmid-plastid-and-mitochondrion-from-refseq" title="Permanent link">:)</a></h3>
<p>Extra repositories from RefSeq release not included as default databases. <a href="https://www.ncbi.nlm.nih.gov/refseq/">Website</a>.</p>
<pre><code class="language-bash"># Download sequence files
wget -A genomic.fna.gz -m -nd --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/plasmid/&quot;
wget -A genomic.fna.gz -m -nd --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/plastid/&quot;
wget -A genomic.fna.gz -m -nd --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/mitochondrion/&quot;

ganon build-custom --input plasmid.* plastid.* mitochondrion.* --db-prefix ppm --level species --threads 8 --input-target sequence
</code></pre>
<h3 id="univec-univec_core">UniVec, UniVec_core<a class="headerlink" href="#univec-univec_core" title="Permanent link">:)</a></h3>
<p>"UniVec is a non-redundant database of sequences commonly attached to cDNA or genomic DNA during the cloning process." <a href="https://ftp.ncbi.nlm.nih.gov/pub/UniVec/README.uv">Website</a>. Useful to screen for vector and linker/adapter contamination. UniVec_core is a sub-set of the UniVec selected to reduce the false positive hits from real biological sources.</p>
<pre><code class="language-bash"># UniVec
wget -O &quot;UniVec.fasta&quot; --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec&quot; 
echo -e &quot;UniVec.fasta\tUniVec\t81077&quot; &gt; UniVec_ganon_input_file.tsv
ganon build-custom --input-file UniVec_ganon_input_file.tsv --db-prefix UniVec --level leaves --threads 8 --skip-genome-size

# UniVec_Core
wget -O &quot;UniVec_Core.fasta&quot; --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec_Core&quot; 
echo -e &quot;UniVec_Core.fasta\tUniVec_Core\t81077&quot; &gt; UniVec_Core_ganon_input_file.tsv
ganon build-custom --input-file UniVec_Core_ganon_input_file.tsv --db-prefix UniVec_Core --level leaves --threads 8 --skip-genome-size
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All UniVec entries in the examples are categorized as <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=81077">Artificial Sequence</a> (NCBI txid:81077). Some are completely artificial but others may be derived from real biological sources. More information in this <a href="https://ftp.ncbi.nlm.nih.gov/pub/UniVec/README.vector.origins">link</a>.</p>
</div>
<h3 id="mgnify-genome-catalogues-mags">MGnify genome catalogues (MAGs)<a class="headerlink" href="#mgnify-genome-catalogues-mags" title="Permanent link">:)</a></h3>
<p>"Genome catalogues are biome-specific collections of metagenomic-assembled and isolate genomes". <a href="https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/">Article</a>/<a href="https://www.ebi.ac.uk/metagenomics/browse/genomes">Website</a>/<a href="https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/">FTP</a>.</p>
<p>Currently available genome catalogues (2024-02-09): <code>chicken-gut</code> <code>cow-rumen</code> <code>honeybee-gut</code> <code>human-gut</code> <code>human-oral</code> <code>human-vaginal</code> <code>marine</code> <code>mouse-gut</code> <code>non-model-fish-gut</code> <code>pig-gut</code> <code>zebrafish-fecal</code></p>
<p><em>List currently available entries <code>curl --silent --list-only ftp://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/</code></em></p>
<p>Example on how to download and build the <code>human-oral</code> catalog:</p>
<pre><code class="language-bash"># Download metadata
wget &quot;https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-oral/v1.0.1/genomes-all_metadata.tsv&quot;

# Download sequence files with 12 threads
tail -n+2 genomes-all_metadata.tsv | cut -f 1,20 | xargs -P 12 -n2 sh -c 'curl --silent ${1}| gzip -d | sed -e &quot;1,/##FASTA/ d&quot; | gzip &gt; ${0}.fna.gz'

# Generate ganon input file
tail -n+2 genomes-all_metadata.tsv | cut -f 1,15  | tr ';' '\t' | awk -F&quot;\t&quot; '{tax=&quot;1&quot;;for(i=NF;i&gt;1;i--){if(length($i)&gt;3){tax=$i;break;}};print $1&quot;.fna.gz\t&quot;$1&quot;\t&quot;tax}' &gt; ganon_input_file.tsv

# Build ganon database
ganon build-custom --input-file ganon_input_file.tsv --db-prefix mgnify_human_oral_v101 --taxonomy gtdb --level leaves --threads 8
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>MGnify genomes catalogues will be build with GTDB taxonomy.</p>
</div>
<h3 id="pathogen-detection-fda-argos">Pathogen detection FDA-ARGOS<a class="headerlink" href="#pathogen-detection-fda-argos" title="Permanent link">:)</a></h3>
<p>A collection of &gt;1400 "microbes that include biothreat microorganisms, common clinical pathogens and closely related species". <a href="https://www.ncbi.nlm.nih.gov/bioproject/231221">Article</a>/<a href="https://www.ncbi.nlm.nih.gov/bioproject/231221">Website</a>/<a href="https://www.ncbi.nlm.nih.gov/bioproject/231221">BioProject</a>.</p>
<pre><code class="language-bash"># Download sequence files
wget https://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt
grep &quot;strain=FDAARGOS&quot; assembly_summary_refseq.txt &gt; fdaargos_assembly_summary.txt
genome_updater.sh -e fdaargos_assembly_summary.txt -f &quot;genomic.fna.gz&quot; -o download -m -t 12

# Build ganon database
ganon build-custom --input download/ --input-recursive --db-prefix fdaargos --ncbi-file-info download/assembly_summary.txt --level assembly --threads 32
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The example above uses <a href="https://github.com/pirovc/genome_updater">genome_updater</a> to download files</p>
</div>
<h3 id="blast-databases-nt-env_nt-nt_prok">BLAST databases (nt env_nt nt_prok ...)<a class="headerlink" href="#blast-databases-nt-env_nt-nt_prok" title="Permanent link">:)</a></h3>
<p>BLAST databases. <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">Website</a>/<a href="https://ftp.ncbi.nlm.nih.gov/blast/db/">FTP</a>.</p>
<p>Current available nucleotide databases (2024-02-09): <code>16S_ribosomal_RNA</code> <code>18S_fungal_sequences</code> <code>28S_fungal_sequences</code> <code>Betacoronavirus</code> <code>env_nt</code> <code>human_genome</code> <code>ITS_eukaryote_sequences</code> <code>ITS_RefSeq_Fungi</code> <code>LSU_eukaryote_rRNA</code> <code>LSU_prokaryote_rRNA</code> <code>mito</code> <code>mouse_genome</code> <code>nt</code> <code>nt_euk</code> <code>nt_others</code> <code>nt_prok</code> <code>nt_viruses</code> <code>patnt</code> <code>pdbnt</code> <code>ref_euk_rep_genomes</code> <code>ref_prok_rep_genomes</code> <code>refseq_rna</code> <code>refseq_select_rna</code> <code>ref_viroids_rep_genomes</code> <code>ref_viruses_rep_genomes</code> <code>SSU_eukaryote_rRNA</code> <code>tsa_nt</code></p>
<p><em>List currently available entries <code>curl --silent --list-only ftp://ftp.ncbi.nlm.nih.gov/blast/db/ | grep "nucl-metadata.json" | sed 's/-nucl-metadata.json/, /g' | sort</code></em></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Some BLAST databases are very big and may require extreme computational resources to build. You may need to use some <a href="../default_databases/#reducing-database-size">reduction strategies</a>.</p>
</div>
<p>The example shows how to <strong>download</strong>, <strong>parse</strong> and <strong>build</strong> a ganon database from BLAST database files. It does so by splitting the database into taxonomic specific files, to speed-up the build process:</p>
<pre><code class="language-bash"># Define BLAST db
db=&quot;16S_ribosomal_RNA&quot;
threads=8

# Download BLAST db - re-run this command many times until all finish (no more output)
curl --silent --list-only ftp://ftp.ncbi.nlm.nih.gov/blast/db/ | grep &quot;^${db}\..*tar.gz$&quot; | xargs -P ${threads:-1} -I{} wget --continue -nd --quiet --show-progress &quot;https://ftp.ncbi.nlm.nih.gov/blast/db/{}&quot;

# OPTIONAL Download and check MD5
wget -O - -nd --quiet --show-progress &quot;ftp://ftp.ncbi.nlm.nih.gov/blast/db/${db}\.*tar.gz.md5&quot; &gt; &quot;${db}.md5&quot;
find -name &quot;${db}.*tar.gz&quot; -type f -printf '%P\n' | xargs -P ${threads:-1} -I{} md5sum {} &gt; &quot;${db}_downloaded.md5&quot;
diff -sy &lt;(sort -k 2,2 &quot;${db}.md5&quot;) &lt;(sort -k 2,2 &quot;${db}_downloaded.md5&quot;)  # Should print &quot;Files /dev/fd/xx and /dev/fd/xx are identical&quot;

# Extract BLAST db files, if successful, remove .tar.gz
find -name &quot;${db}.*tar.gz&quot; -type f -printf '%P\n' | xargs -P ${threads} -I{} sh -c 'gzip -dc {} | tar --overwrite -vxf - &amp;&amp; rm {}' &gt; &quot;${db}_extracted_files.txt&quot;

# Create folder to write sequence files (split into 10 sub-folders)
seq 0 9 | xargs -i mkdir -p &quot;${db}&quot;/{}

# This command extracts sequences from the blastdb and writes them into taxid specific files
# It also generates the --input-file for ganon with the fields: filepath &lt;tab&gt; file &lt;tab&gt; taxid
blastdbcmd -entry all -db &quot;${db}&quot; -outfmt &quot;%a %T %s&quot; | \
awk -v db=&quot;$(realpath ${db})&quot; '{file=db&quot;/&quot;substr($2,1,1)&quot;/&quot;$2&quot;.fna&quot;; print &quot;&gt;&quot;$1&quot;\n&quot;$3 &gt;&gt; file; print file&quot;\t&quot;$2&quot;.fna\t&quot;$2}' | \
sort | uniq &gt; &quot;${db}_ganon_input_file.tsv&quot;

# Build ganon database
ganon build-custom --input-file &quot;${db}_ganon_input_file.tsv&quot; --db-prefix &quot;${db}&quot; --threads ${threads} --level leaves

# Delete extracted files and auxiliary files
cat &quot;${db}_extracted_files.txt&quot; | xargs rm
rm &quot;${db}_extracted_files.txt&quot; &quot;${db}.md5&quot; &quot;${db}_downloaded.md5&quot;
# Delete sequences and input_file
rm -rf &quot;${db}&quot; &quot;${db}_ganon_input_file.tsv&quot;
</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code>blastdbcmd</code> is a command from <a href="https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST+ software suite</a> (tested version 2.14.0) and should be installed separately.</p>
</div>
<h3 id="files-from-genome_updater">Files from genome_updater<a class="headerlink" href="#files-from-genome_updater" title="Permanent link">:)</a></h3>
<p>To create a ganon database from files previously downloaded with <a href="https://github.com/pirovc/genome_updater">genome_updater</a>:</p>
<pre><code class="language-bash">ganon build-custom --input output_folder_genome_updater/version/ --input-recursive --db-prefix mydb --ncbi-file-info  output_folder_genome_updater/assembly_summary.txt --level assembly --threads 32
</code></pre>
<h2 id="parameter-details">Parameter details<a class="headerlink" href="#parameter-details" title="Permanent link">:)</a></h2>
<h3 id="false-positive-and-size-max-fp-filter-size">False positive and size (--max-fp, --filter-size)<a class="headerlink" href="#false-positive-and-size-max-fp-filter-size" title="Permanent link">:)</a></h3>
<p>ganon indices are based on bloom filters and can have false positive matches. This can be controlled with <code>--max-fp</code> parameter. The lower the <code>--max-fp</code>, the less chances of false positives matches on classification, but the larger the database size will be. For example, with <code>--max-fp 0.01</code> the database will be build so any target (defined by <code>--level</code>) will have 1 in a 100 change of reporting a false k-mer match. <a href="../classification/#false-positive-of-a-query-fpr-query">The false positive of the query</a> (all k-mers of a read) will be way lower, but directly affected by this value.</p>
<p>Alternatively, one can set a specific size for the final index with <code>--filter-size</code>. When using this option, please observe the theoretic false positive of the index reported at the end of the building process.</p>
<h3 id="minimizers-window-size-kmer-size">minimizers (--window-size, --kmer-size)<a class="headerlink" href="#minimizers-window-size-kmer-size" title="Permanent link">:)</a></h3>
<p>in <code>ganon build</code>, when <code>--window-size</code> &gt; <code>--kmer-size</code> minimizers are used. That means that for a every window, a single k-mer will be selected. It produces smaller database files and requires substantially less memory overall. It may increase building times but will have a huge benefit for classification times. Sensitivity and precision can be reduced by small margins. If <code>--window-size</code> = <code>--kmer-size</code>, all k-mers are going to be used to build the database.</p>
<h3 id="target-file-or-sequence-input-target">Target file or sequence (--input-target)<a class="headerlink" href="#target-file-or-sequence-input-target" title="Permanent link">:)</a></h3>
<p>This is a parameter that defines how ganon will parse your input files:
 - <code>--input-target file</code> (default) will consider every file provided with <code>--input</code> a single unit  (e.g. multi-fasta files are considered one input, sequence headers ignored).
 - <code>--input-target sequence</code> will use every sequence as a unit. For this, ganon will first decompose every sequence in the input files provided with <code>--input</code> into a separated file. This will take longer and use more disk space.</p>
<p><code>--input-target file</code> is the default behavior and most efficient way to build databases. <code>--input-target sequence</code> should only be used when the input sequences are not separated by file (e.g. a single big FASTA file) or when classification at sequence level is desired.</p>
<h3 id="build-level-level">Build level (--level)<a class="headerlink" href="#build-level-level" title="Permanent link">:)</a></h3>
<p>The <code>--level</code> parameter defines the max. depth of the database for classification. This parameter is relevant because the <code>--max-fp</code> is going to be guaranteed at the <code>--level</code> chosen. </p>
<p>In <code>ganon build</code> the default value is <code>species</code>. In <code>ganon build-custom</code> the level will be the same as <code>--input-target</code>, meaning that classification will be done either at <code>file</code> or <code>sequence</code> level.</p>
<p>Alternatively, <code>--level assembly</code> will link the file or sequence target information with assembly accessions retrieved from NCBI. <code>--level leaves</code> or <code>--level species</code> (or <code>genus</code>, <code>family</code>, ...) will link the targets with taxonomic information and prune the tree at the chosen level. <code>--level custom</code> will use specialization (4th col.) defined in the <code>--input-file</code>.</p>
<h3 id="genome-sizes-genome-size-files">Genome sizes (--genome-size-files)<a class="headerlink" href="#genome-sizes-genome-size-files" title="Permanent link">:)</a></h3>
<p>Ganon will automatically download auxiliary files to define an approximate genome size for each entry in the taxonomic tree. For <code>--taxonomy ncbi</code> the <a href="https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/">species_genome_size.txt.gz</a> is used. For <code>--taxonomy gtdb</code> the <a href="https://data.gtdb.ecogenomic.org/releases/latest/">*_metadata.tar.gz</a> files are used. Those files can be directly provided with the <code>--genome-size-files</code> argument.</p>
<p>Genome sizes of parent nodes are calculated as the average of the respective children nodes. Other nodes without direct assigned genome sizes will use the closest parent with a pre-calculated genome size. The genome sizes are stored in the <a href="#buildupdate">ganon database</a>.</p>
<h3 id="retrieving-info-ncbi-sequence-info-ncbi-file-info">Retrieving info (--ncbi-sequence-info, --ncbi-file-info)<a class="headerlink" href="#retrieving-info-ncbi-sequence-info-ncbi-file-info" title="Permanent link">:)</a></h3>
<p>Further taxonomy and assembly linking information has to be collected to properly build the database. <code>--ncbi-sequence-info</code> and <code>--ncbi-file-info</code> allow customizations on this step.</p>
<p>When <code>--input-target sequence</code>, <code>--ncbi-sequence-info</code> argument allows the use of NCBI e-utils webservices (<code>eutils</code>) or downloads accession2taxid files to extract target information (options <code>nucl_gb</code> <code>nucl_wgs</code> <code>nucl_est</code> <code>nucl_gss</code> <code>pdb</code> <code>prot</code> <code>dead_nucl</code> <code>dead_wgs</code> <code>dead_prot</code>). By default, ganon uses <code>eutils</code> up-to 50000 input sequences, otherwise it downloads <code>nucl_gb</code> <code>nucl_wgs</code> from https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/. Previously downloaded files can be directly provided with this argument.</p>
<p>When <code>--input-target file</code>, <code>--ncbi-file-info</code> uses <code>assembly_summary.txt</code> from https://ftp.ncbi.nlm.nih.gov/genomes/ to extract target information (options <code>refseq</code> <code>genbank</code> <code>refseq_historical</code> <code>genbank_historical</code>. Previously downloaded files can be directly provided with this argument.</p>
<p>If you are using outdated, removed or inactive assembly or sequence files and accessions from NCBI, make sure to include <code>dead_nucl</code> <code>dead_wgs</code> for <code>--ncbi-sequence-info</code> or <code>refseq_historical</code> <code>genbank_historical</code> for <code>--ncbi-file-info</code>. <code>eutils</code> option does not work with outdated accessions.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../default_databases/" class="btn btn-neutral float-left" title="Databases (ganon build)"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../classification/" class="btn btn-neutral float-right" title="Classification (ganon classify)">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../default_databases/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../classification/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../js/extra.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
